keps/sig-windows/20190418-windows-security-context.md

```diff
@@ -0,0 +1,151 @@
+
+- GMSACredentialSpecName: A string specifying the name of a GMSA credential spec custom resource
+- GMSACredentialSpec: A string specifying the full credential spec JSON string associated with a GMSA credential spec
+- RunAsUserName: A string specifying the user name in Windows to run the entrypoint of the container
+
+More fields may be added to the `WindowsSecurityOptions` struct as desired in the future.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The `WindowsSecurityOptions` struct will be defined as follows in pkg/apis/core/types.go
+
+```
+// WindowsSecurityOptions contain Windows-specific options and credentials.
+type WindowsSecurityOptions struct {
+	// GMSACredentialSpecName is the name of the GMSA credential spec to use.
+	// +optional
+	GMSACredentialSpecName string
+
+	// GMSACredentialSpec is where the GMSA admission webhook
+	// (https://github.com/kubernetes-sigs/windows-gmsa) inlines the contents of the
+	// GMSA credential spec named by the `GmsaCredentialSpecName` field.
+	// +optional
+	GMSACredentialSpec string
+
+	// RunAsUserName is the local user context used to log in to the container
+	// +optional
+	RunAsUserName string
+}
+```
+
+Field `WindowsOptions *WindowsSecurityOptions` will be added to `SecurityContext` and `PodSecurityContext` structs.
+
+#### Specification of both GMSA credspec and RunAsUserName
+
+Note that both GMSA credspec and RunAsUserName may be specified. Specification of one field is not mutually exclusive with the other. RunAsUserName governs the local user identity used to log into the container. This is decoupled from the GMSA domain identity used to interact with network resources. To use GMSA identity, processes in the container should run as "Local System" or "Network Service" users. However Kubernetes won't enforce any rules around specification of these fields. For further details, please refer [here](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/manage-serviceaccounts#configuring-your-application-to-use-the-gmsa)
+
+#### Changes in kubelet
+
+An effective value for a field in a container's `SecurityContext.WindowsOptions` will be determined by calling `DetermineEffectiveSecurityContext`. If a value is specified for a field in the pod's `PodSecurityContext.WindowsOptions`, `DetermineEffectiveSecurityContext` will apply it to the corresponding field in a container's `SecurityContext.WindowsOptions` unless that field is already set with a different value in the container's `SecurityContext.WindowsOptions`.
+
+### Risks and Mitigations
+
+None
+
+## Design Details
+
+### Test Plan
+
+Unit tests will be added around `DetermineEffectiveSecurityContext` to ensure for each field in `WindowsSecurityOptions` the values in a container's `SecurityContext.WindowsOptions` can override values of the fields in the pod's `PodSecurityContext.WindowsOptions`.
+
```

Comment:
We can also include a E2E test for this that will be labelled `[sig-windows]` and included in the existing test passes.